### ISSUE TITLE
AWS Certified Machine Learning Preparation Notes

### ISSUE THREAD

**Issue by Joldnine** [Joldnine/joldnine.github.io#42]
Recently in 2020 December, I just passed AWS Certified Machine Learning exam with a score of 931/1000. Here is the preparation notes for the exam.

# ML Domain Knowledge
## Feature Engineering
1. Imputing missing data
Mean/median replacement, median is better when got outliers
Dropping
KNN, deep learning, regression (MICE, multiple imputation by chained equations)
Just get more data

2. Handling unbalanced data
Oversampling
Undersampling
SMOTE, synthetic minority over-sampling technique
Adjusting thresholds

3. Handling outlier
What is outlier
Data points that lie more than certain (one) standard deviation from the mean.
 How to resolve
Remove outliers when understand them.

4. Techniques of feature engineering
Binning
Transforming with functions
Encoding
Scaling/normalization
Shuffling

5. TF-IDF Score
Term Frequency and Inverse Document Frequency: figure out what terms are most relevant for a document
Term frequency: how often a word occurs in a document
Document frequency: how often a word in an entire set of documents --> common words that appear everywhere
Relevancy of a word to a document: TF/DF == TF*IDF (IDF = 1/DF) --> how often the word appears in a document over how often it appears everywhere, ie. how important and unique this word is for this document
An extension: use uni-grams, bi-grams, tri-grams, n-grams

### Deep Learning
1. Activation functions
Define the output of node givens its input signals
Linear activation function: no backpropagation
Binary step function: no multiple classfication, not good for calculus
Sigmoid, logistic, tanH (for RNN)
Rectified Linear Unit (ReLU)
Leaky ReLU
Parametric ReLU, negative slope is learned via backpropagation
Exponential Linear Unit (ELU)
Swish for really deep NN
Softmax: final output layer, prob of each class

2. CNN
What are they for:
Images, translation, sentence classification, sentiment analysis
Feature-location invariant
How do they work:
Local receptive fields

3. How to fix vanishing gradient problem
Multi-level heirarchy
Long short-term memory
Residual networks
Better choice of activation function, ReLU

4. Batch Size
Small —> out local minima, better

5. Regularization, dropout

## Metrics
Recall (TP/(TP+FN))
Precision (TP/(TP+FP))
F1 2PR/(P+R)
ROC, AUC
RMSE

# AWS
## AWS Data Engineering
Glue: Data Catalog, Crawler, ETL (Spark), 
Catalog, crawler

Athena
Serverless query from S3, columnar data formats (Parquet, Apache ORC) faster perf

QuickSight
Visualization, ML Insights, adhoc

Streaming
Kinesis: source —> Streams —> Analytics —> Firehose —> s3/redshift

Step Functions
Workflow

Batch
Resource and schedule

EMRFS
Access S3 as if it were HDFS

## AWS Embedded Algorithms
XGBoost
Subsample, Eta prevent overfitting

The Amazon SageMaker Latent Dirichlet Allocation (LDA) algorithm
an unsupervised learning algorithm that attempts to describe a set of observations as a mixture of distinct categories.
Latent Dirichlet Allocation, unsupervised, topic modelling, use CPU

Random Cut Forest
Anomaly detection

Neural Topic Model
Classify or summarize documents based on topics, Unsupervised

Factorization Machines
Sparse data, click prediction, item recom, recordIO float32

IP Insights
Unsupervised, suspicious behavior from IP addresses, CSV only

Seq2seq 
Tokens must be integers

BlazingText
Text classification, __label__ 
Continuous bag of words, order does not matter

SageMaker RL
Distributed training

Amazon Comprehend
NLP and text analytics, sentiment

Amazon Transcribe
Speech

Amazon Polly
Text to speech

Amazon Lex
Alexa chatbot engine

Amazon Forecast
Time series

Amazon Kendra
Search with natural language

DeepAR
Time series, RNN

## Sagemaker
SageMaker Neo
Compiling models using TensorFlow and other frameworks to edge devices such as Nvidia Jetson

Sagemaker production variants
Production new model switching inside Sagemaker

Any bucket with “sagemaker” in the name is accessible with the default SageMaker notebook role.

Sagemaker inference 
Container response to 8080 for /invocations and /ping
Artefacts in tar format

Pipe mode
Streams data directly to container, improve perf

Most Amazon SageMaker algorithms work best when you use the optimised protobuf recordIO format for the training data.
